pkg/policy/rules.go

```diff
@@ -359,13 +360,21 @@ loop:
 // added to epIDSet. Note that epIDSet can be shared across goroutines!
 // Returns whether the endpoint was selected by one of the rules, or if the
 // endpoint is nil.
-func (rules ruleSlice) updateEndpointsCaches(ep Endpoint, epIDSet *IDSet) bool {
+func (rules ruleSlice) updateEndpointsCaches(ep Endpoint, epIDSet *IDSet) (bool, error) {
 	if ep == nil {
-		return true
+		return false, fmt.Errorf("cannot update caches in rules because endpoint is nil")
 	}
 	id := ep.GetID16()
+	if err := ep.RLockAlive(); err != nil {
+		return false, fmt.Errorf("cannnot update caches in rules for endpoint %d because it is being deleted: %s", id, err)
+	}
+	defer ep.RUnlock()
```

Comment:
Do we want to release the lock right after obtaining the security identity? If yes, we might want to put `securityIdentity` retrieval into inline closure func with deferred unlock.

Reply:
There's another case here which is that when an endpoint is deleted causing an identity to no longer be used locally, the identitymanager will call out into the policy package to flush that identity out from all of the rule caches here. We want to make sure that this `updateEndpointsCaches()` case finishes all of its work before that other side is triggered, otherwise there's a race where:

1. The policy is updated, causing a call to `updateEndpointsCaches()`
1. We get the Identity here
1. The endpoint is removed
1. The endpoint removal causes the policy repository to get updated to remove this Identity from these caches (note: This does not take the policy repo write lock right now because it's only changing the rule caches)
1. Finally, this code path adds the Identity into the rule caches

This would result in Identities in the rule caches to exist even though the identity isn't locally used by any endpoint. We would never clean this up.

(Now, maybe there's another solution to this particular problem but at least for now it's probably safer to keep this Endpoint RLock to block out endpoint deletion until the caches are updated from the policy update side here).

Reply:
Actually, might be worth double-checking (4) above. Surely we *should* be taking a readlock to go over the repository, which would automatically solve this already..

Reply:
@joestringer I'm not 100% sure that taking the `RLock` of the repository there guarantees the race might not occur, because the policy repository lock is not always held when `updateEndpointsCaches` due to goroutines being spawned to perform the work of updating the caches.

See: 

https://github.com/cilium/cilium/blob/9765d3adffd1554793aeee123a80b23e17b5b6ed/daemon/policy.go#L378-L380

 We have a WaitGroup passed into the funtion from `policyAdd` and `policyDelete` which we could wait on *before* releasing the repository mutex?  I did this intentionally when I wrote the selective regeneration work since I didn't want to hold the lock of the repository itself, since the rule content itself wasn't being changed, rather the caches themselves, which have their own mutex. I think holding the endpoint mutex for the duration of this is OK, as keeping one endpoint locked longer vs. the policy repository is better, since the repository is a global structure. 

All of these mutexes are making my head hurt 😢

Reply:
@ianvernon Hmm, interesting point.... is that safe? Shouldn't we at least hold the RLock while iterating rules in the repository during cache update, to ensure that two policy changes in quick succession don't end up causing havoc with the the cache update goroutines?

Reply:
@joestringer what sort of havoc were you envisioning? The policy addition logic itself is protected by the queueing logic, so we should be insulated from such a case. A policy computation for an endpoint could occur while we are updating the caches, though, but that's OK because the content of the rules themselves is not changing.

My goal here was to reduce contention on the policy repository lock since it's used across the entire system. I do think that this logic is far too complicated, though. Too many means of serialization makes it impossible to reason about how things work =/

Reply:
My question is with the following order of events:

1. Policy update A comes in, is queued into `RepositoryChangeQueue`
1. Policy update B comes in, is queued into `RepositoryChangeQueue`
1. Take A off the `RepositoryChangeQueue`, grab the mutex
1. Make policy changes for A
1. Trigger selection cache updates for A
1. Release the mutex
1. Add events to handle regerenation for A to `RuleReactionQueue`
1. Take B off the `RepositoryChangeQueue`, grab the mutex
1. Start processing changes for B
1. ...

How do you know that (8) blocks on the selection update completing in goroutines triggered from (5)?

I see a policySelectionWG, but that seems at a glance to be passed down to the rule reaction queue to ensure that the processing associated with (7) doesn't happen until after (5)'s work is complete. However AFAICT it doesn't say anything about ensuring that (5)'s work completes before (9) starts manipulating the repository.

Reply:
Discussed offline. The operation `UpdateRulesEndpointsCaches`, which updates the caches, is performed upon slices of pointers to rules from the policy repository. Not the repository itself. So, we do not have to worry about the policy repository being manipulated while these slices are being analyzed.